src/Badges.sol

```diff
@@ -121,26 +207,52 @@ contract Badges is
 
   function tokenURI(uint256 _tokenId) external view virtual override returns (string memory) {
     require(exists(_tokenId), "tokenURI: token doesn't exist");
+
     return tokenURIs[_tokenId];
   }
 
-  function unequip(uint256 _tokenId) external virtual override {
-    require(owners[_tokenId] != address(0), "unequip: token doesn't exist");
+  function unequip(uint256 _tokenId) external virtual override tokenExists(_tokenId) {
     require(msg.sender == owners[_tokenId], "unequip: sender must be owner");
+
     uint256 voucherHashId = voucherHashIds[_tokenId];
     usedHashes.unset(voucherHashId);
     burn(_tokenId);
   }
 
   function balanceOf(address _owner) external view virtual override returns (uint256) {
     require(_owner != address(0), "balanceOf: address zero is not a valid owner_");
+
     return balances[_owner];
   }
 
-  function ownerOf(uint256 _tokenId) external view virtual override returns (address) {
-    address owner_ = owners[_tokenId];
-    require(owner_ != address(0), "ownerOf: token doesn't exist");
-    return owner_;
+  function ownerOf(uint256 _tokenId) external view virtual override tokenExists(_tokenId) returns (address) {
+    return owners[_tokenId];
+  }
+
+  function revokeBadge(uint256 _raftTokenId, uint256 _badgeId) external senderIsRaftOwner(_raftTokenId) {
+    revokedBadges[_badgeId] = true;
+  }
+
+  function reinstateBadge(uint256 _raftTokenId, uint256 _badgeId)
+    external
+    tokenExists(_badgeId)
+    senderIsRaftOwner(_raftTokenId)
+  {
+    delete revokedBadges[_badgeId];
+  }
+
+  function updateExpiration(
+    uint256 _raftTokenId,
+    uint256 _timestamp,
+    uint256 _badgeId
+  ) external tokenExists(_badgeId) senderIsRaftOwner(_raftTokenId) {
```

Comment:
Hmm.. `senderIsRaftOwner` I think this is incorrect. I think this needs to be an internal function without scoping it to raftOwner. No?

Reply:
This function isn't called from "mint". Maybe we don't need it in this PR but the idea is that if a badge expires there should be a mechanism for the expiration date to get extended.

I imagine that the only person who would be able to update the expiration of a badge would be the raft owner.

Reply:
Ah. To support renewal of expired badges? 🤔 Yeah, in that case it would be an external call by Spec Creator / Raft Owner. But I gues the scope of spec creator comes in when we work on the "Jupiter Permissions" project. 

Can you check with Emily if this is absolutely needed, please?